Fix instability in sorting replay frames

frenzibyte opened the pull request:
Resolves #12651

frenzibyte committed:
Fix replay frames sort instability

frenzibyte committed:
Return back removed using

ekrctb commented on osu.Game/Rulesets/Replays/FramedReplayInputHandler.cs:
```diff
@@ -97,7 +98,7 @@ protected FramedReplayInputHandler(Replay replay)
             // TODO: This replay frame ordering should be enforced on the Replay type.
             // Currently, the ordering can be broken if the frames are added after this construction.
-            replay.Frames.Sort((x, y) => x.Time.CompareTo(y.Time));
+            replay.Frames = replay.Frames.OrderBy(f => f.Time).ToList();
```
Note that if a ruleset stores a reference to the `replay.Frames`, it changes behavior (mutation to the list no longer applied).

ekrctb reviewed (dismissed):
I left a note. As I wrote in the TODO comment, I think more encapsulation is needed for the `Replay` type, like the consumer shouldn't have access to the direct reference to the `List<ReplayFrame>`.
But this is fine as a quick fix.

peppy commented:
This 100% needs a test.

frenzibyte committed:
Add test coverage

frenzibyte commented:
I'm not 100% sure on the test I wrote as I've never wrote any sort stability test before, but I think it's good enough, kinda?

bdach commented on osu.Game.Tests/NonVisual/FramedReplayInputHandlerTest.cs:
```diff
@@ -278,6 +280,38 @@ public void TestMultipleFramesSameTime()
+
+            int data = 0;
+
+            // 1. add a range of frames in which some of them have the constant time 5000, all without any "data".
+            // 2. randomize the frames.
+            // 3. assign "data" to each frame in ascending order.
+            replay.Frames.AddRange(Enumerable.Range(1, 250).Select(i =>
+            {
+                if (RNG.NextBool())
+                    return new TestReplayFrame(repeating_time, true);
+                else
+                    return new TestReplayFrame(i * 1000, true);
+            }).OrderBy(_ => RNG.Next()).Select(f => new TestReplayFrame(f.Time, true, ++data)));
```
I'm very not sold on this. The RNG is unseeded, so this test case is essentially non-deterministic. If it ever fails, nobody will ever know why because the seed used will be the system clock value at the time of initialising RNG, and good luck recovering that. And yes, if you seed it then it is possible that an unstable sort, purely by chance, doesn't shuffle the order and the test will be a false positive. But (1) the more same-time frames you have, the lesser the possibility, semi-theoretically speaking, and (2) I'd probably rather risk that than have a test that is undebuggable because you can never literally get the same values again.

Heck, I'd probably even hardcode a simple 4-frame example for this. This isn't theoretically exhaustively testable anyways. The data space is too large.

frenzibyte replied:
Ah, now I realize how bad this test case I wrote is. I've tried hardcoding a 4-frame example for this but got a false positive (testing before change), I'll try adding more and see.

frenzibyte replied:
Got it failing on 20-frames, should be good enough now rather than whatever bullshit that was here.

frenzibyte committed:
Avoid randomizing and overestimating logic with simple hardcoding

Not sure what was in my mind while I was pushing that..

frenzibyte committed:
Use LINQ select for data assigning for simplicity

To avoid having to read through all of frames and ensure nothing is failing there

bdach committed:
Simplify test case further

bdach committed:
Reword & clarify comments

bdach reviewed (approved):
Made some more simplifications, reworded comments a touch. Seems OK now, but probably wise to have this go through a second pair of eyes.

caulmseh commented:
Tried to autoplay centipede before and after this update and osu now hits it with 100% accuracy still though with frame drops
